hub/api/dataset.py

```diff
@@ -21,68 +19,77 @@
             memory_cache_size (int): The size of the memory cache to be used in MB.
             local_cache_size (int): The size of the local filesystem cache to be used in MB.
-            storage (StorageProvider, optional): The storage provider used to access
-                the data stored by this dataset. If this is specified, the path given is ignored.
+            creds (dict, optional): A dictionary containing credentials used to access the dataset at the path.
+                This takes precedence over credentials present in the environment. Currently only works with s3 paths.
+                It supports 'aws_access_key_id', 'aws_secret_access_key', 'aws_session_token', 'endpoint_url' and 'region' as keys.
+            storage (StorageProvider, optional): The storage provider used to access the dataset.
+                Use this if you want to specify the storage provider object manually instead of using a path to generate it.
 
         Raises:
             ValueError: If an existing local path is given, it must be a directory.
-            UserWarning: Both path and storage should not be given.
-            UnsupportedModeError: If mode is not any of the options listed above.
+            ImproperDatasetInitialization: Exactly one argument out of 'path' and 'storage' needs to be specified.
+                This is raised if none of them are specified or more than one are specifed.
+            InvalidHubPathException: If a Hub cloud path (path starting with hub://) is specified and it isn't of the form hub://username/datasetname.
+            AuthorizationException: If a Hub cloud path (path starting with hub://) is specified and the user doesn't have access to the dataset.
+            PathNotEmptyException: If the path to the dataset doesn't contain a Hub dataset and is also not empty.
         """
-        if not mode in SUPPORTED_MODES:
-            raise UnsupportedModeError(mode)
+        if creds is None:
+            creds = {}
+        base_storage = get_storage_provider(path, storage, read_only, creds)
 
-        if storage is not None and path:
-            warnings.warn(
-                "Dataset should not be constructed with both storage and path. Ignoring path and using storage."
-            )
-        elif storage is not None and hasattr(storage, "root"):
-            # Extract the path for printing, if path not given
-            self.path = storage.root  # type: ignore
+        # done instead of directly assigning read_only as backend might return return read_only permissions
+        if hasattr(base_storage, "read_only") and base_storage.read_only:
+            self.read_only = True
         else:
-            self.path = path  # Used for printing, if given
+            self.read_only = False
 
-        base_storage = storage or storage_provider_from_path(path)
+        self.path = path or get_path_from_storage(base_storage)  # Used for printing
```

Comment:
Not sure what the comment means.

Reply:
Just saw the `__str__` below. Can you change the word `printing` to something that is more clear?